### PR TITLE
jwt_authn: make "issuer" field in JwtProvider as optional

### PR DESCRIPTION
**Pull request by qiwzhang**
Signed-off-by: Wayne Zhang <qiwzhang@google.com>

To fix https://github.com/envoyproxy/envoy/issues/12377

According to [rfc](https://tools.ietf.org/html/rfc7519#section-4.1.1), "iss" field should be optional.  But "issuer" in JwtProvider filter_config is marked as required.  

This PR changes it to be optional.  Its usage implications are following:
* If a Jwt has "iss" field and the "issuer" field in JwtProvider config is specified, they have to match. In all other cases,  the JWT "iss" field is not checked.
* If allow_missing or allow_missing_or_fail is used,  the Jwt "iss" field is used to find JwtProvider to verify the signature. If the Jwt doesn't have "iss" field, the first JwtProvider without "issuer" is used.

Commit Message:
Additional Description:
Risk Level: None as new features are added to allow empty issuer in the JwtProvider
Testing: Added unit-tests
Docs Changes: None
Release Notes: Yes

### CODE REVIEW

**Comment by qiwzhang** [comment]
@yangminzhu @nareddyt could you help to review this code?  Thanks

File: api/envoy/extensions/filters/http/jwt_authn/v3/config.proto [hunk 1]
```diff
@@ -59,11 +59,19 @@ message JwtProvider {
 
   // Specify the `principal <https://tools.ietf.org/html/rfc7519#section-4.1.1>`_ that issued
   // the JWT, usually a URL or an email address.
+  // It is optional. If specified, it has to match the *iss* field in JWT.
+  // Its usage implications are following:
+  // * If a JWT has *iss* field and this field is specified, they have to match.
+  // * If a JWT doesn't have *iss* field,  not to check with this field.
```
Comment (yangminzhu) [inline api/envoy/extensions/filters/http/jwt_authn/v3/config.proto:65]: so envoy only checks that the `iss` field in JWT is matched with the `issuer` field here when they both exist, and skips the check in all other cases? can we rephrase a little bit, I find the current comment a bit hard to follow.

And regarding the JwtProvider lookup in line 68, isn't the `provider` defined in a map, how do you which one is the first one?

File: test/extensions/filters/http/jwt_authn/authenticator_test.cc [hunk 2]
```diff
@@ -174,8 +174,60 @@ TEST_F(AuthenticatorTest, TestJwtWithNonExistKid) {
...
+      }));
+
+  // Token with other issuer is OK since "issuer" is not specified in the provider.
+  Http::TestRequestHeaderMapImpl headers{
+      {"Authorization", "Bearer " + std::string(OtherGoodToken)}};
+  expectVerifyStatus(Status::Ok, headers);
+}
+
+// Jwt "iss" is "other.com", "issuer" in JwtProvider is not specified,
+// authenticator doesn't have a valid provider. The verification is OK.
```
Comment (yangminzhu) [inline test/extensions/filters/http/jwt_authn/authenticator_test.cc:208]: why is the verification ok when the authenticator doesn't have a valid provider? shouldn't it fail in this case.

Also in what case will the authenticator not have a valid provider?

File: api/envoy/extensions/filters/http/jwt_authn/v3/config.proto (hunk 1 above)
Comment (qiwzhang) [reply api/envoy/extensions/filters/http/jwt_authn/v3/config.proto:67]: Will rephase the comment to make it clear.

Good point.  "the first" is not deterministic with the proto map.  
Since "allow_missing" used with Jwt without "iss" are rare cases, I will not try to design a way to make "the first" deterministic.  I will just document it.  If anybody wants to use it,  they have to make sure there is not one provider without issuer specified.

File: test/extensions/filters/http/jwt_authn/authenticator_test.cc (hunk 2 above)
Comment (qiwzhang) [reply test/extensions/filters/http/jwt_authn/authenticator_test.cc:208]: I added comment in source/extensions/filters/http/jwt_authn/authenticator.cc line 163 when provider is not valid.
I will copy it here to readability.

Comment (qiwzhang) [reply test/extensions/filters/http/jwt_authn/authenticator_test.cc:208]: If a Provider doesn't have "issuer" specified,  it can be used for any Jwt in the invalid provider case, which is in "allow_missing" or "allow_failed" case.

File: source/extensions/filters/http/jwt_authn/extractor.cc [hunk 3]
```diff
@@ -22,6 +22,36 @@ namespace HttpFilters {
...
+      return true;
+    }
+    return specified_issuers_.find(jwt_issuer) != specified_issuers_.end();
+  }
+
+private:
+  // If true, all issuers are ok
+  bool allow_all_{false};
+  // Only these specified issuers are allowed.
+  absl::node_hash_set<std::string> specified_issuers_;
```
Comment (nareddyt) [inline source/extensions/filters/http/jwt_authn/extractor.cc:52]: Nit: `absl::flat_hash_set` would be a better choice

go/hashmaps#recommendation-2

File: source/extensions/filters/http/jwt_authn/authenticator.cc [hunk 4]
```diff
@@ -166,8 +159,15 @@ void AuthenticatorImpl::startVerify() {
   // Check the issuer is configured or not.
   jwks_data_ = provider_ ? jwks_cache_.findByProvider(provider_.value())
                          : jwks_cache_.findByIssuer(jwt_->iss_);
-  // isIssuerSpecified() check already make sure the issuer is in the cache.
-  ASSERT(jwks_data_ != nullptr);
+  // When `provider` is valid, findByProvider should never return nullptr.
+  // Only when `allow_missing` or `allow_failed` is used, `provider` is invalid,
```
Comment (nareddyt) [inline source/extensions/filters/http/jwt_authn/authenticator.cc:163]: I'm fairly confused about this

> Only when `allow_missing` or `allow_failed` is used, `provider` is invalid

The code here suggests `provider_` can be set even when `allow_missing` is true.

https://github.com/envoyproxy/envoy/blob/fdd1693b49b5bf0d0cea9bed839c3b98f3b1b6e8/source/extensions/filters/http/jwt_authn/authenticator.cc#L105

Can you clarify why `provider_` should be nullptr?

Comment (qiwzhang) [reply source/extensions/filters/http/jwt_authn/authenticator.cc:163]: When allow_missing or allow_failed is used, the authenticator is designed to work on all providers, not just for one particular provider.  Hence provider is invalid.

File: source/extensions/filters/http/jwt_authn/extractor.cc (hunk 3 above)
Comment (qiwzhang) [reply source/extensions/filters/http/jwt_authn/extractor.cc]: Done

**Review by nareddyt — APPROVED** [review]

**Comment by qiwzhang** [comment]
@lizan @htuch and envoyproxy/api-shepherds ping

File: api/envoy/extensions/filters/http/jwt_authn/v3/config.proto [hunk 5]
```diff
@@ -60,10 +60,25 @@ message JwtProvider {
   // Specify the `principal <https://tools.ietf.org/html/rfc7519#section-4.1.1>`_ that issued
   // the JWT, usually a URL or an email address.
   //
+  // It is optional. If specified, it has to match the *iss* field in JWT.
+  //
+  // If a JWT has *iss* field and this field is specified, they have to match, otherwise the
```
Comment (htuch) [inline api/envoy/extensions/filters/http/jwt_authn/v3/config.proto:65]: For the non-empty issuer case, have any semantics actually changed, or is this documentation change purely about clarifying existing behavior?

Comment (qiwzhang) [reply api/envoy/extensions/filters/http/jwt_authn/v3/config.proto:65]: Not semantics change for non-empty issuer case. These comment just clarify the existing behavior

File: api/envoy/extensions/filters/http/jwt_authn/v3/config.proto [hunk 6]
```diff
@@ -60,10 +60,25 @@ message JwtProvider {
...
+  // JWT *iss* field is not checked.
+  //
+  // If *allow_missing* or *allow_missing_or_failed* is used, the JWT *iss* field is used to find
```
Comment (htuch) [inline api/envoy/extensions/filters/http/jwt_authn/v3/config.proto:68]: If the JWT is missing (with `allow_missing`), where do you get the JWT *iss* field from?

Comment (qiwzhang) [reply api/envoy/extensions/filters/http/jwt_authn/v3/config.proto:68]: This is for the case where if there is a JWT, how to handle it.

Comment (htuch) [reply api/envoy/extensions/filters/http/jwt_authn/v3/config.proto:68]: Maybe I'm a bit confused then; why does `allow_missing` have special semantics when the JWT is not missing?

Comment (qiwzhang) [reply api/envoy/extensions/filters/http/jwt_authn/v3/config.proto:68]: with "allow_missing",  if there is a JWT token in the request,  we still have to verify it and if verification fails,  the request is rejected.  But if JWT is missing,  it is OK, the request is proceeded.   This is used for "optional" JWT verification usage.

Comment (qiwzhang) [reply api/envoy/extensions/filters/http/jwt_authn/v3/config.proto:68]: I guess my previous comment did not answer your question.  Why semantics is different in `allow_missing` case?

For supporting backward compatibility, 'allowing_missing` and `allow_failed` is implemented differently than the other normal cases.   In other normal cases,  it is against a specific `provider`, not need to use "iss" to find the "provider".  But in the former case,  it is against all providers,  we have to use "iss" from JWT to find the provider to verify.  That is why the semantic different.

Comment (htuch) [reply api/envoy/extensions/filters/http/jwt_authn/v3/config.proto:68]: Makes sense. Could you add a `[#comment: ]` explaining this for future folks who are trying to understand? I appreciate your patience here!

**Review by htuch — COMMENTED** [review]
/lgtm api

File: api/envoy/extensions/filters/http/jwt_authn/v3/config.proto (hunk 6 above)
Comment (qiwzhang) [reply api/envoy/extensions/filters/http/jwt_authn/v3/config.proto]: Done

**Comment by qiwzhang** [comment]
Sorry,  did a force-push.  Hope to fix a test failure unrelated to this change.

**Comment by qiwzhang** [comment]
It had passed all tests, but after added a few comment in the proto, some tests failed.

**Comment by qiwzhang** [comment]
Hi @htuch I just did a force-push to cause a retest.  there is not new changes.  Could you help to review it again?  Thanks

File: api/envoy/extensions/filters/http/jwt_authn/v3/config.proto [hunk 7]
```diff
@@ -60,10 +60,24 @@ message JwtProvider {
...
   //
+  // It is optional. If specified, it has to match the *iss* field in JWT.
+  //
+  // If a JWT has *iss* field and this field is specified, they have to match, otherwise the
+  // JWT *iss* field is not checked.
+  //
+  // Note: *JwtRequirement* :ref:`allow_missing <envoy_v3_api_field_extensions.filters.http.jwt_authn.v3.JwtRequirement.allow_missing>`
+  // and :ref:`allow_missing_or_failed <envoy_v3_api_field_extensions.filters.http.jwt_authn.v3.JwtRequirement.allow_missing_or_failed>`
+  // are implemented differently than other *JwtRequirements*. Hence the usage of this field
+  // is different as followings if *allow_missing* or *allow_missing_or_failed* is used:
```
Comment (mattklein123) [inline api/envoy/extensions/filters/http/jwt_authn/v3/config.proto:71]:
```suggestion
  // is different as follows if *allow_missing* or *allow_missing_or_failed* is used:
```

File: docs/root/version_history/current.rst [hunk 8]
```diff
@@ -75,6 +75,7 @@ New Features
 * http: added frame flood and abuse checks to the upstream HTTP/2 codec. This check is off by default and can be enabled by setting the `envoy.reloadable_features.upstream_http2_flood_checks` runtime key to true.
 * http: clusters now support selecting HTTP/1 or HTTP/2 based on ALPN, configurable via :ref:`alpn_config <envoy_v3_api_field_extensions.upstreams.http.v3.HttpProtocolOptions.auto_config>` in the :ref:`http_protocol_options <envoy_v3_api_msg_extensions.upstreams.http.v3.HttpProtocolOptions>` message.
 * jwt_authn: added support for :ref:`per-route config <envoy_v3_api_msg_extensions.filters.http.jwt_authn.v3.PerRouteConfig>`.
+* jwt_authn: changed config field :ref:`issuer <envoy_v3_api_field_extensions.filters.http.jwt_authn.v3.JwtProvider.issuer>` to be optional.
```
Comment (mattklein123) [inline docs/root/version_history/current.rst:78]: Can you write a bit more about why this was done?

**Review by mattklein123 — CHANGES_REQUESTED** [review]
Some small API/doc comments. Thank you!

/wait

File: api/envoy/extensions/filters/http/jwt_authn/v3/config.proto (hunk 7 above)
Comment (qiwzhang) [reply api/envoy/extensions/filters/http/jwt_authn/v3/config.proto]: Done

File: docs/root/version_history/current.rst (hunk 8 above)
Comment (qiwzhang) [reply docs/root/version_history/current.rst]: Done

File: docs/root/version_history/current.rst [hunk 9]
```diff
@@ -75,6 +75,7 @@ New Features
 * http: added frame flood and abuse checks to the upstream HTTP/2 codec. This check is off by default and can be enabled by setting the `envoy.reloadable_features.upstream_http2_flood_checks` runtime key to true.
 * http: clusters now support selecting HTTP/1 or HTTP/2 based on ALPN, configurable via :ref:`alpn_config <envoy_v3_api_field_extensions.upstreams.http.v3.HttpProtocolOptions.auto_config>` in the :ref:`http_protocol_options <envoy_v3_api_msg_extensions.upstreams.http.v3.HttpProtocolOptions>` message.
 * jwt_authn: added support for :ref:`per-route config <envoy_v3_api_msg_extensions.filters.http.jwt_authn.v3.PerRouteConfig>`.
+* jwt_authn: changed config field :ref:`issuer <envoy_v3_api_field_extensions.filters.http.jwt_authn.v3.JwtProvider.issuer>` to be optional to comply with JWT [RFC](https://tools.ietf.org/html/rfc7519#section-4.1.1) requirements.
```
Comment (mattklein123) [inline docs/root/version_history/current.rst:78]: Unfortunately markdown external links do not work, can you switch to RST style and check the docs?

/wait

Comment (qiwzhang) [reply docs/root/version_history/current.rst]: Done

File: docs/root/version_history/current.rst [hunk 10]
```diff
@@ -75,7 +75,7 @@ New Features
 * http: added frame flood and abuse checks to the upstream HTTP/2 codec. This check is off by default and can be enabled by setting the `envoy.reloadable_features.upstream_http2_flood_checks` runtime key to true.
 * http: clusters now support selecting HTTP/1 or HTTP/2 based on ALPN, configurable via :ref:`alpn_config <envoy_v3_api_field_extensions.upstreams.http.v3.HttpProtocolOptions.auto_config>` in the :ref:`http_protocol_options <envoy_v3_api_msg_extensions.upstreams.http.v3.HttpProtocolOptions>` message.
 * jwt_authn: added support for :ref:`per-route config <envoy_v3_api_msg_extensions.filters.http.jwt_authn.v3.PerRouteConfig>`.
-* jwt_authn: changed config field :ref:`issuer <envoy_v3_api_field_extensions.filters.http.jwt_authn.v3.JwtProvider.issuer>` to be optional to comply with JWT [RFC](https://tools.ietf.org/html/rfc7519#section-4.1.1) requirements.
+* jwt_authn: changed config field :ref:`issuer <envoy_v3_api_field_extensions.filters.http.jwt_authn.v3.JwtProvider.issuer>` to be optional to comply with JWT `RFC <https://tools.ietf.org/html/rfc7519#section-4.1.1>_` requirements.
```
Comment (mattklein123) [inline docs/root/version_history/current.rst:78]: This is still not correct. Please look at the rendered docs. (undescore at the end)

/wait

Comment (qiwzhang) [reply docs/root/version_history/current.rst]: Done. Thanks.

**Review by mattklein123 — APPROVED** [review]
